File: NeetCode/Stack Algorithms/ReversePolishNotation.java
import java.util.Stack;

public class ReversePolishNotation {
    public static int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();    

        for(String s : tokens) {                                        //! Time Complexity = O(n)
            if(s.matches("-?\\d+")) {
                stack.push(Integer.parseInt(s));
            } else if(s.charAt(0) == '+' || s.charAt(0) == '-' || s.charAt(0) == '*' || s.charAt(0) == '/') {
                int a = stack.pop();
                int b = stack.pop();
                int res = calc(a, b, s.charAt(0));
                stack.push(res);
            } else {
                throw new IllegalArgumentException("Invalid Input:" + s);
            }
        }

        return stack.pop();
    }                                                                   //! Total Time Complexity = O(n)

    public static int calc(int a, int b, char operator) {
        if(operator == '+') {
            return b+a;
        } else if(operator == '-') {
            return b-a;
        } else if(operator == '*') {
            return b*a;
        } else {
            return b/a;
        }
    }

    public static void main(String[] args) {
        // String[] tokens = {"2", "1", "+", "3", "*"};
        // String[] tokens = {"4", "13", "5", "/", "+"};
        String[] tokens = {"10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"};

        System.out.println(evalRPN(tokens));
    }   
}

File: NeetCode/Stack Algorithms/ReversePolishNotation.cs

/**
 * Reverse Polish Notation
 *
 * You are given an array of strings tokens that represents a valid arithmetic expression in Reverse Polish Notation.
 * Return the integer that represents the evaluation of the expression.
 * The operands may be integers or the results of other operations.
 * The operators include '+', '-', '*', and '/'.
 * Assume that division between integers always truncates toward zero.
 *
 * Test Case:
 * Input: tokens = ["1","2","+","3","*","4","-"]
 * Output: 5
 * Explanation: ((1 + 2) * 3) - 4 = 5
 * 
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 *
*/

public class Solution {
    public int EvalRPN(string[] tokens) {
        Stack<int> stack = new Stack<int>();
        int a = 0, b = 0;
        foreach(string c in tokens) {
            switch(c)
            {
                case "+":
                    a = stack.Pop();
                    b = stack.Pop();
                    stack.Push(a+b);
                    break;
                case "-":
                    a = stack.Pop();
                    b = stack.Pop();
                    stack.Push(b-a);
                    break;
                case "*":
                    a = stack.Pop();
                    b = stack.Pop();
                    stack.Push(a*b);
                    break;
                case "/":
                    a = stack.Pop();
                    b = stack.Pop();
                    stack.Push(b/a);
                    break;
                default:
                    stack.Push(Convert.ToInt32(c));
                    break;
            }
        }

        return stack.Pop();
    }
}

